src/scraper/http.cpp

```diff
@@ -51,16 +65,92 @@ namespace
 
         return curl;
     }
+
+    struct progress {
+      timetype lastruntime;
+      CURL *curl;
+    };
+
+    static int newerprogress_callback(void *ptr, curl_off_t downtotal, curl_off_t downnow, curl_off_t uptotal, curl_off_t uplnow)
+    {
+        struct progress *pg = (struct progress*)ptr;
+        CURL *curl = pg->curl;
+
+        // Thread inturrupting
+        try
+        {
+            boost::this_thread::interruption_point();
+            // Set this once.
+            if (DownloadStatus.SnapshotDownloadSize == 0)
+                DownloadStatus.SnapshotDownloadSize = downtotal;
+
+            timetype currenttime = 0;
+            curl_easy_getinfo(curl, timeopt, &currenttime);
+
+            // Update every 1 second
+            if ((currenttime - pg->lastruntime) >= progressinterval)
+            {
+                pg->lastruntime = currenttime;
+
+                curl_off_t speed;
+                CURLcode result;
+
+                result = curl_easy_getinfo(curl, CURLINFO_SPEED_DOWNLOAD_T, &speed);
+
+                // Download speed update
+                if (result == CURLE_OK)
+                {
+
+#if LIBCURL_VERSION_NUM >= 0x073700
+                    if (speed > 0)
+                        DownloadStatus.SnapshotDownloadSpeed = speed;
+
+                    else {
+                        DownloadStatus.SnapshotDownloadSpeed = 0;
+
+                    }
+#else
+                    // Not supported by libcurl
+                    DownloadStatus.SnapshotDownloadSpeed = -1;
+#endif
+                    if (DownloadStatus.SnapshotDownloadSize > 0 && (downnow > 0))
+                    {
+                        DownloadStatus.SnapshotDownloadProgress = ((downnow / (double)DownloadStatus.SnapshotDownloadSize) * 100);
+                        DownloadStatus.SnapshotDownloadAmount = downnow;
+                    }
+                }
+            }
+        }
+
+        catch (boost::thread_interrupted&)
+        {
+            return 1;
+        }
+
+        return 0;
+    };
+
+#if LIBCURL_VERSION_NUM < 0x072000
+    static int olderprogress_callback(void *ptr, double downtotal, double downnow, double uptotal, double upnow)
+    {
+        return newerprogress_callback(ptr, (curl_off_t)downtotal, (curl_off_t)downnow, (curl_off_t)uptotal, (curl_off_t)uptotal, (curl_off_t)upnow);
+    };
+
+#endif
 }
 
-Http::Http()
+Http::Http(bool Scoped)
 {
-    curl_global_init(CURL_GLOBAL_ALL);
+    fScoped = Scoped;
+
+    if (fScoped)
+        curl_global_init(CURL_GLOBAL_ALL);
```

Comment:
thats original I just made it so that my snapshot download for file does not use the scoped curl since it'll seg fault because the progress monitor. This allows it to be used with all aspects exception download of the file. but yes @jamescowens touch base here. I know allowing this also causes seg fault since I already clean up curl in the code in download of snapshot file

Reply:
Hmm....  we should have @denravonska to comment on this. I think it is ok for this PR.

Reply:
I think I was mistaken--there's a potential race condition for those curl functions for scraper nodes when the update checker is enabled. The update checker may try to call one of these functions while the scraper thread is download project stats or vice-versa. 

It could be enough to set `disableupdatecheck=1` on scraper nodes. Even better, we force the update checker to remain disabled when running with `-scraper`. 

The best solution may be to statically initialize/cleanup curl once for the life of the program.

Reply:
Er, nevermind--the update checker runs with `fScoped = false`.

Reply:
Well, the update checker will eventually call `curl_easy_init()`, which does invoke `curl_global_init()`. That should be fine, but the scraper thread will call `curl_global_cleanup()` when the `Http` objects are destroyed, so there is a race for the resources managed by curl.

Reply:
> If you did not already call curl_global_init, curl_easy_init does it automatically. This may be lethal in multi-threaded cases, since curl_global_init is not thread-safe, and it may result in resource problems because there is no corresponding cleanup. 

https://curl.haxx.se/libcurl/c/curl_easy_init.html

Reply:
Can we go to a single instance, like I did with the scraper log?

Reply:
Actually, why don't we call curl_global_init early in init.cpp and be done with it.

Reply:
Looks like we need to create a custom method for initializing the curl_global_init, and then eliminate the curl_global_init and curl_global_cleanup on the constructor and destructor. Seems like all that is needed is the curl_easy_cleanup for normal object destruction. @cyrossignol?

Reply:
How about something like... https://github.com/jamescowens/Gridcoin-Research/commit/1e5f6871ccad95c39a8cc728bd5df898997ed95f

Maybe a pretty dumb attempt. Haven't tried to run it yet.

Reply:
How about something like https://github.com/jamescowens/Gridcoin-Research/commit/48928eb8f824ff27adc42c0df3cd134d1debccb6.

I did this against the development branch. Admittedly pretty dumb and may not work. Haven't tried to run it yet...

Reply:
That looks like it will work... 

I was thinking of something more like: 

```cpp 
class Http 
{
    ...
private: 
    struct CurlLifecycle 
    {
        CurlLifecycle() { curl_global_init(); }
        ~CurlLifecycle() { curl_global_cleanup(); }
    };

    static CurlLifecycle curl_lifecycle;
};
```

This will automatically call the functions and static initialization guarantees thread-safety. It behaves as recommended by the curl documentation:

> The basic rule for constructing a program that uses libcurl is this: Call curl_global_init, with a CURL_GLOBAL_ALL argument, immediately after the program starts, while it is still only one thread and before it uses libcurl at all. Call curl_global_cleanup immediately before the program exits, when the program is again only one thread and after its last use of libcurl.

Reply:
That is much simpler than my idea, of course. @iFoggz we should do what @cyrossignol suggested. This makes the constructor and destructor of individual http objects trivial. It also means you should be able to remove the workaround.

Reply:
The advantage to my method is that curl only takes up memory space if there is one or more active objects.